### DISCUSSION TITLE
Where are the bookmarks stored?

### DISCUSSION THREAD

**Discussion by keunes** [organicmaps/organicmaps#517] [Q&A] [answered]
Hello,

I wanted to create a custom folder sync with my Nextcloud, but I don't see the app in which the OrganicMaps bookmarks are stored. I do have a folder called "OMapsData" which could be the one (given the name), but the 'bookmarks' folder doesn't contain the new bookmark list I just created.

I also have a folder Android > data > app.organicmaps, but there's no 'bookmarks' subfolder.

**Comment by biodranik** [comment]
@rtsisyk please help :)

**Comment by amo-f77** [comment, marked as answer]
I would love to know this too. As I like to manually copy/backup my kml/kmz files.
Normally I would look in settings.ini to find the location, but I can't find that either.

Android > data > app.organicmaps is the most logical place for them.

**Reply by amo-f77** [reply]
Found the location using data from and android backup for the app.
`adb backup -noapk app.organicmaps`

A quick extract  and the settings.ini shows the location of the kml files to be
`/data/user/0/app.organicmaps/files/bookmarks`

Which is not accessible to me, as I have not rooted my tablet.

It would be nice if the settings.ini and the kml files were in a normal user accessible location on the tablet/phone.
I'm sort of happy knowing that I can grab them all with adb.

**Reply by keunes** [reply]
Thanks @amo-f77 for finding this out and reporting back. Agree it would be nice if these files were in a user-accesible location. Should we create an issue for it?

**Reply by amo-f77** [reply]
@keunes, it's probably not issue worthy. There may be good reasons for the location and bookmarks work as intended.
At most, I'd say this is more an "Ideas" item.

**Reply by rtsisyk** [reply]
> It would be nice if the settings.ini and the kml files were in a normal user accessible location on the tablet/phone.
> I'm sort of happy knowing that I can grab them all with adb.

Unfortunately, such location doesn't exist on Android 11+ anymore. Android 11 is 45-50% of market share already and growing.

Google is forcing all developers to move to so called ["scoped storage"](https://developer.android.com/about/versions/11/privacy/storage).

We tried multiple options:

1. Old `WRITE_EXTERNAL_STORAGE` permission to write `/sdcard` - ignored completely, access denied.
2. `getExternalFilesDir()` - returns `/sdcard/Android/data/app.organicmaps` (sdcard), it works, but this directory is inaccessible from other apps including file managers.
3. `getFilesDir()` - returns `/data/user/0/app.organicmaps/` (internal storage) - current.
4. StorageAccessFramework to write `/sdcard/Documents/OrganicMaps' - this is a compeletely weird. You can write to a folder, but the application can read only its files. In other words, if you copy a bookmark file by using a file manager, it will exist in the folder, but the app will not see it.
5. New MANAGE_EXTERNAL_STORAGE to write to arbitrary directory - requires a special approval from Google, designed for filemanagers. I'm not sure that privacy-focused map  actually needs permission to read all files.

Probably it is easier to add a button "Export all bookmarks into a folder" when deal with new Google limitations.

**Reply by rtsisyk** [reply]
BTW, bookmarks should be included into the system backup. Could you please check that it works on your device?

**Reply by keunes** [reply]
> it's probably not issue worthy. There may be good reasons for the location and bookmarks work as intended.
At most, I'd say this is more an "Ideas" item.

Right, I'm still not used to this 'discussions' thing :p In my mind there's just different issue types.

Thanks a lot @rtsisyk for the explanation!

What do you mean with 'system backup'? Like, a backup of the whole of Android?

I don't know, and I guess that wouldn't serve my use-case (syncing between my phone & Nextcloud).

I think I'll create an 'Idea' discussion for direct integration/synchronization with Nextcloud - reckon that under Android 11 that's the only solution.

**Reply by amo-f77** [reply]
Thanks for the detailed explanation @rtsisyk. Trust google to take out the simplicity.

I like the idea of "export all kml to a folder".

Not a big fan of google app data backup/restore, it only saves the latest copy, which could be a corrupt one. Plus I can't see the files inside the backup, only the app name, size and date. I rarely use it anymore. I prefer backups I can control, keep versions and see contents.

To answer your question @rtsisyk. Yes, App backup seems to work by restoring settings and kml files. (Android 7.1.1)
Long version: I re-enabled app backup and triggered it with adb. A backup for organicmaps appeared on my google drive. I stopped the app, cleared data & cache, re-ran organicmaps and confirmed it was blank. Uninstalled organimaps and then reinstalled it. Apart from losing my map, which is easily downloaded, everything was there.

My other reason for access to the directory of kml files is to run a quick script on my tablet to strip all `<mwm:icon>   </mwm:icon>` lines out. So that all my bookmarks are shown with stars and not a mix of stars and icon symbols. See Ideas  https://github.com/organicmaps/organicmaps/discussions/740.  Like the really old mapswithmepro of 2013/14.

**Reply by keunes** [reply]
I just created a request in the 'Ideas' section for [synchronisation with Nextcloud Maps](https://github.com/organicmaps/organicmaps/discussions/776).

**Reply by HimmyWolf** [reply]
Hello!

 Sorry for reviving this old thread. I came through this thread through a google search about the locstion of rhe Organic Maps boomarks data. I've possibly lost my android phone to water damage. It may come down to data recovery from the motherboard or other sort of data recovery, if the phone can't be fixed. For recovering photos and other auch files it won't be an issue, but I am worried about the big list of bookmarks that I had saved on my travela through Organic Maps. The last KMZ file export and back up that I did was several months ago so a lot of important bookmarks would be lost. Is there any way to access those bookmarks through data recovery - using a laptop for example to view the recovered files or for example copying them to an SD card which would be inserted into a smart phone that has the same model and same Android version? Or is there no other way to salvage those bookmarks other than to recover the phone's function and open the application itself?

Thank you in advance
Kaspar

**Comment by alejandrogallo** [comment]
when is the bookmark exporting to a folder feature is going to be implemented, I'm really locked with hundreds of bookmarks that I want to back up asap.

**Reply by biodranik** [reply]
Do you have many bookmark lists? You can export each list manually now and save the KMZ file where you want.

**Reply by brunetton** [reply]
Same here; I've 20 bookmarks lists; and exporting one by one makes another problem appears: file explorer apps I've tested refuse to handle the case where file already exists in output directory and makes it impossible to replace existant file. It makes it impossible to **update previously exported file**

I've tested quite a lot of file managers before finding one that actually makes it possible. Here are some examples:
- [Material files](https://github.com/zhanghai/MaterialFiles) -> refuse to replace existing file
- [Solid explorer](https://play.google.com/store/apps/details?id=pl.solidexplorer2) -> creates ` (1).kmz` files (example: existing my_places.kmz; the app creates another file named `my_places (1).kmz`; then `my_places (2).kmz` etc). This is a problem as I'm using Synthings to sync those bookmarks and it already have an internal versions management

Finally I managed to replace existing files with [Simple File Manager](https://github.com/SimpleMobileTools/Simple-File-Manager); but the whole process of exporting my 20 lists is very time consuming and error prone.

## Workaround suggested

I'd love to be able to export all files at once in a pre-defined directory (in preferences I guess)


Thanks a lot

**Reply by brunetton** [reply]
Answering to myself: there is already an issue exactly for that: https://github.com/organicmaps/organicmaps/issues/995

**Comment by tch123** [comment]
Can you set the default list of point favorites, I found that now only the newly created list can be added

**Reply by biodranik** [reply]
Bookmarks are added to the last used list. If you add a bookmark, then change its list, after that all new bookmarks will be added to that list. If it doesn't work like that, then it's a bug, please let us know.
